### PR TITLE
funding: adding to topology after fundinglocked, private channels, chanAnn after 6 confs

### PR DESCRIPTION
**Pull request by Crypt-iQ**
This PR addresses issue #306 and the rest of #305. My other [PR](https://github.com/lightningnetwork/lnd/pull/319) addresses the first part of decoupling. It is dependent on the other PR.

This commit introduces some new interdependent functionality. As soon as the fundingLocked message is sent, the channel is immediately added to the ChannelRouter's internal topology. Additionally, there is now an option for private channels when calling the openchannel rpc call. The --private flag initiates a
private channel. There are tests in fundingmanager_test.go that assert the privacy guarantees of non-broadcasting. Finally, channels are now only broadcasted to the greater network after six confirmations on the funding transaction has been reached.

### CODE REVIEW

File: fundingmanager.go [hunk 1]
```diff
@@ -1435,20 +1584,191 @@ func (f *fundingManager) sendFundingLockedAndAnnounceChannel(
...
+		LastUpdate:	time.Unix(int64(chanUpdateAnn.Timestamp), 0),
+		Flags:		chanUpdateAnn.Flags,
+		TimeLockDelta:	chanUpdateAnn.TimeLockDelta,
+		MinHTLC:	chanUpdateAnn.HtlcMinimumMsat,
+		FeeBaseMSat:	lnwire.MilliSatoshi(chanUpdateAnn.BaseFee),
+		FeeProportionalMillionths:lnwire.MilliSatoshi(chanUpdateAnn.FeeRate),
+	}
+
+	// Add ChannelEdgePolicy to ChannelRouter's topology
+	err = f.cfg.UpdateEdge(update); if err != nil {
```
Comment (jimpo) [inline fundingmanager.go]: I feel like it's preferable to use the normal `SendAnnouncement` code path to process the channel update. Only thing is that is is unclear whether the local node will send the channel_update to the remote peer since `AnnounceSignatures` is not sent, but it should do that regardless because it is not necessary to make a channel public.

File: peer.go [hunk 2]
```diff
@@ -666,8 +666,89 @@ out:
 			isChanUpdate = true
 			targetChan = msg.ChanID
 
-		case *lnwire.ChannelUpdate,
```
Comment (jimpo) [inline peer.go]: Can you explain why this was changed and ProcessRemoteAnnouncement cannot be used?

Comment (Crypt-iQ) [reply peer.go]: In `AnnounceSignatures`, two nodes exchange `ChannelEdgePolicy` and update their peer's directed edges in their own channel graph. Now, per issue #306, it is possible that they don't want their channel announced to the greater network. This means not sending an `AnnounceSignatures` message. But there must be some way of exchanging `ChannelUpdate` messages, so I altered `peer.go` to include functionality to determine when it is acceptable to pass it to the gossiper, and when it is acceptable to NOT (to preserve privacy).

Comment (Crypt-iQ) [reply peer.go]: I was also told by @Roasbeef to not use flags in the `gossiper.go` to gate such behavior... I know that `peer.go` isn't pretty, but I wasn't sure what else to do.

Comment (Crypt-iQ) [reply peer.go]: I think you are right that I can use `ProcessRemoteAnnouncement` ... but I'm less sure about using it for `AddEdge` & `UpdateEdge`

Comment (Crypt-iQ) [reply peer.go]: I suppose I can do this in `ProcessRemoteAnnouncement`:

1) If we receive a `ChannelAnnouncement`, `isRemote` is false and `ChannelEdgeInfo` is not returned from `GetChannelByID`, we don't announce & we call `Router.AddEdge`.
2) If we receive a `ChannelUpdate`, `isRemote` is false and no `ChannelEdgePolicy` is returned from `GetChannelByID`, we don't announce & we call `Router.UpdateEdge`. We can just not include an authentication proof here and it won't be announced.

Let me think about `AnnounceSignatures` and get back to you.

Comment (Crypt-iQ) [reply peer.go]: We can also just include the `AuthProof` in step 2) above.
This would mean that the `ChannelUpdate` would be sent to the remote peer and using similar logic to the logic you commented about would mean the `AnnounceSignatures` branch wouldn't need to be called. Let me know what you think.

Comment (Crypt-iQ) [reply peer.go]: My above comment is a bit wrong, we shouldn't include the `AuthProof` in step 2).  Instead if the code follows the branch that doesn't announce `ChannelUpdate`, this code branch should also send the `ChannelUpdate` message directly to the peer so there is no announcement happening.

Comment (Crypt-iQ) [reply peer.go]: Alright I reverted the changes on `peer.go` and the relevant code is now in `gossiper.go`.  Let me know what you think.  It works as before, but this is better I believe so thanks for the comments!

**Comment by Crypt-iQ** [comment]
Note: the Travis build fails because `lnd_test.go` uses 1 mined block to open the channel, instead of 6.

**Comment by Crypt-iQ** [comment]
@jimpo @halseth It seems that my original `private` flag approach was the right idea (though not the rest of my approach). Because of this line [here](https://github.com/lightningnetwork/lnd/blob/master/discovery/gossiper.go#L455), I should be adding a `private` flag to stored channels so that when the `retransmitTimer` ticks, it won't broadcast our channel.

**Comment by Crypt-iQ** [comment]
In my most recent commit, I have fixed the integration tests and `golint` complaints.

File: discovery/gossiper.go [hunk 3]
```diff
@@ -511,7 +511,7 @@ func (d *AuthenticatedGossiper) retransmitStaleChannels() error {
 		// If it's been a full day since we've re-broadcasted the
 		// channel, add the channel to the set of edges we need to
 		// update.
-		if timeElapsed >= broadcastInterval {
+		if timeElapsed >= broadcastInterval && info.AuthProof != nil {
```
Comment (Roasbeef) [inline discovery/gossiper.go]: Minor nit: the above comment should also be updated to outline that we check the auth proof, as we shouldn't broadcast channels which we don't have the full proofs for, or have not be _announced_ to the network. In the case of the latter case, it may be a privacy leak.

File: discovery/gossiper.go [hunk 4]
```diff
@@ -871,6 +871,26 @@ func (d *AuthenticatedGossiper) processNetworkAnnouncement(nMsg *networkMsg) []l
 			return nil
 		}
 
+		// If this is a local ChannelUpdate announcement, send it to
+		// our peer.
+		if !nMsg.isRemote {
```
Comment (Roasbeef) [inline discovery/gossiper.go]: Shouldn't think be _only_ done in the case that this is the first time we're learning of this channel?

File: fundingmanager.go [hunk 5]
```diff
@@ -322,6 +333,12 @@ var (
 	// of being opened.
 	channelOpeningStateBucket = []byte("channelOpeningState")
 
+	// openChanAnnPrefBucket is the database bucket used to store each
+	// channel's announcement preference signalled by the LSB of
+	// channel_flags of the open_channel message in the funding workflow.
+	// It only stores open channels' announcement preferences.
+	openChanAnnPrefBucket = []byte("open_chan_ann")
```
Comment (Roasbeef) [inline fundingmanager.go]: Minor nit: the bucket name should be modified to match the styling of the bucket above (so change to camel case).

File: fundingmanager.go [hunk 6]
```diff
@@ -454,11 +472,6 @@ func (f *fundingManager) Start() error {
 		f.newChanBarriers[chanID] = make(chan struct{})
 		f.barrierMtx.Unlock()
 
-		// Set up a localDiscoverySignals to make sure we finish sending
-		// our own fundingLocked and channel announcements before
-		// processing a received fundingLocked.
-		f.localDiscoverySignals[chanID] = make(chan struct{})
```
Comment (Roasbeef) [inline fundingmanager.go]: Why was this removed?

File: fundingmanager.go [hunk 7]
```diff
@@ -483,11 +496,21 @@ func (f *fundingManager) Start() error {
 
 		case fundingLockedSent:
 			// fundingLocked was sent to peer, but the channel
+			// was not added to the router graph and the channel
 			// announcement was not sent.
 			f.wg.Add(1)
 			go func() {
 				defer f.wg.Done()
 
+				// Retrieve channel announcement preference
```
Comment (Roasbeef) [inline fundingmanager.go]: This block of code is starting to sprawl a bit. It should be extract into a distinct method for readability purposes. This should likely be done in a distinct commit _before_ the new logic is added, as this will make the change easier to review, and will also make the change (move only) explicit in the git history.

File: fundingmanager.go [hunk 8]
```diff
@@ -1582,24 +1733,90 @@ func (f *fundingManager) sendFundingLocked(completeChan *channeldb.OpenChannel,
...
+	}
+
+	// Send ChannelAnnouncement and ChannelUpdate to the gossiper to add
+	// to the Router's topology.
+	if err = f.cfg.SendLocalAnnouncement(ann.chanAnn); err != nil {
+		return fmt.Errorf("error sending private channel "+
+			"announcement: %v", err)
+	}
+	if err = f.cfg.SendLocalAnnouncement(ann.chanUpdateAnn); err != nil {
+		return fmt.Errorf("error sending private channel "+
```
Comment (Roasbeef) [inline fundingmanager.go]: The error message here, and above don't necessarily match the context.

File: fundingmanager.go [hunk 9]
```diff
@@ -1582,24 +1733,90 @@ func (f *fundingManager) sendFundingLocked(completeChan *channeldb.OpenChannel,
...
-// be used.
-func (f *fundingManager) sendChannelAnnouncement(completeChan *channeldb.OpenChannel,
+// addToRouterGraph sends a private ChannelAnnouncement and a private
+// ChannelUpdate to the gossiper so that it is added to the Router's internal
+// graph before the announcement_signatures is sent in
+// annAfterSixConfs. These private announcement messages are NOT
+// broadcasted to the greater network.
+func (f *fundingManager) addToRouterGraph(completeChan *channeldb.OpenChannel,
+	channel *lnwallet.LightningChannel,
+	shortChanID *lnwire.ShortChannelID, private bool) error {
```
Comment (Roasbeef) [inline fundingmanager.go]: Seems the `private` arg isn't used anywhere in this method. As a result, it can safely be removed.

File: fundingmanager.go [hunk 10]
```diff
@@ -497,13 +520,68 @@ func (f *fundingManager) Start() error {
...
+
+				if !private {
+					err = f.annAfterSixConfs(channel,
+						lnChannel, shortChanID)
+					if err != nil {
+						fndgLog.Errorf("error sending channel "+
+							"announcement: %v", err)
+						return
+					}
+				}
```
Comment (Roasbeef) [inline fundingmanager.go]: Seems like the channel state can be deleted at this point if the channel is marked as private.

File: fundingmanager.go [hunk 11]
```diff
@@ -1582,24 +1733,90 @@ func (f *fundingManager) sendFundingLocked(completeChan *channeldb.OpenChannel,
...
+// is sent and the channel is added to the router graph (channelState is
+// 'addedToRouterGraph') and the channel is ready to be used.
+func (f *fundingManager) annAfterSixConfs(completeChan *channeldb.OpenChannel,
 	channel *lnwallet.LightningChannel, shortChanID *lnwire.ShortChannelID) error {
 
-	// TODO(eugene) wait for 6 confirmations here
+	// Register with the ChainNotifier for a notification once the
+	// funding transaction reaches 6 confirmations.
+	txid := completeChan.FundingOutpoint.Hash
+	confNtfn, err := f.cfg.Notifier.RegisterConfirmationsNtfn(&txid, 6,
```
Comment (Roasbeef) [inline fundingmanager.go]: I think this should be modified to actually announce after `max(6, completeChan.NumConfsRequired)`. The `FundingLocked` message is only sent once the channel has reached the required number of confirmations, and the channel announcement (if not private) should _only_ be sent after the `FundingLocked` message has been sent. As a result, there's a causal dependency that we need to ensure we uphold.

File: fundingmanager.go [hunk 12]
```diff
@@ -2216,6 +2422,86 @@ func copyPubKey(pub *btcec.PublicKey) *btcec.PublicKey {
...
+		if err != nil {
+			return err
+		}
+
+		var outpointBytes bytes.Buffer
+		if err = writeOutpoint(&outpointBytes, chanPoint); err != nil {
+			return err
+		}
+
+		scratch := make([]byte, 0)
```
Comment (Roasbeef) [inline fundingmanager.go]: [binary.Write](https://golang.org/pkg/encoding/binary/#Write) can be used here as it now supports writing boolean arguments values.

File: fundingmanager.go [hunk 13]
```diff
@@ -2216,6 +2422,86 @@ func copyPubKey(pub *btcec.PublicKey) *btcec.PublicKey {
...
+			return err
+		}
+
+		value := bucket.Get(outpointBytes.Bytes())
+		if value == nil {
+			return fmt.Errorf("Channel announcement preference " +
+				"not found")
+		}
+
+		if value[0] == 1 {
```
Comment (Roasbeef) [inline fundingmanager.go]: [binary.Read](https://golang.org/pkg/encoding/binary/#Read) can be used here as it now supports reading boolean arguments values.

File: fundingmanager.go (hunk 6 above)
Comment (halseth) [reply fundingmanager.go]: If I'm not mistaken, the `localDiscoverySignal` has the purpose of making sure we have seen the funding tx confirm on-chain before handling an incoming `fundingLocked`. In this case the channels already have their funding txs confirmed, we just need to send the fundinglocked and channel announcements, hence we won't need the `localdiscoverysignal`. This was added by me by mistake earlier, and became apparent here since we now will wait six confirmations before sending the channel announcements, and previously we wouldn't handle the incoming fundinglocked before this was done.

**Comment by halseth** [comment]
PR that continues this one, and addresses the comments: #411

**Comment by Roasbeef** [comment]
Closing as #411 has taken this over.
